test/e2e/run-all.js

```diff
@@ -30,9 +31,44 @@ const getTestPathsForTestDir = async (testDir) => {
   return testPaths;
 };
 
+// Quality Gate Retries
+const RETRIES_FOR_NEW_OR_CHANGED_TESTS = 5;
+let changedOrNewTests;
+
+/**
+ * Runs the quality gate logic to filter and append changed or new tests if present.
+ *
+ * @param {string} fullTestList - List of test paths to be considered.
+ * @returns {string} The updated full test list.
+ */
+async function applyQualityGate(fullTestList) {
+  let qualityGatedList = fullTestList;
+  changedOrNewTests = await filterE2eChangedFiles();
```

Comment:
Nit: Perhaps we could generate this earlier and pass it in, or return it alongside the quality gate list?

With this variable at the module scope, it may be easy to introduce errors later if we try to use it before `applyQualityGate` is called. And if/when we convert this file to TypeScript, it will make the type `| undefined` which is more difficult to deal with. We can avoid both problems by using `cost` to declare the variable, either in `runningOnCircleCI` or in both functions.

Reply:
I've declared it in `runningOnCirclCI` and then passed the `changedOrNewTests` it to `applyQualityGate`. I'm also returning this on the `runningOnCircleCi` since we need it to apply the no-retry flag for the tests that are new/changed